---
title: "Reclustering Radial Glia"
output: html_document
date: "2025-02-04"
---
Libraries

```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE)
{
  library(parallel)
  library(clusterProfiler)
  library(blme)
  library(Seurat)
  library(tidyverse)
  library(tidyr)
  library(lme4)
  library(dplyr)
  library(MASS)
  library(SeuratObject)
  library(Signac)
  library(glmGamPoi)
  library(scran)
  library(parallel)
  library(factoextra)
  library(readxl)
  library(factoextra)
  library(forcats)
  library(ggrepel)
  library(biomaRt)
  library(openxlsx)
  library(emmeans)
  library(CytoTRACE)
  library(ggrepel)
  library(tidyverse)
  library(cowplot)
  library(patchwork)
  library(Polychrome)
  library(scCustomize)

P40 <- createPalette(40, c("#FF0000", "#00FF00", "#0000FF"), range = c(30, 80))
swatch(P40)
names(P40) <- NULL

mean_expression_cluster_plot<- readRDS('Functions/mean_expression_cluster_plot.rds')
prop_cluster_plot<- readRDS( 'Functions/prop_cluster_plot.rds')
define_degs_prop<- readRDS('Functions/define_degs_prop.rds')
mean_expression_cluster_data<- readRDS('Functions/mean_expression_cluster_data.rds')
prop_deg_function.rds<- readRDS('Functions/DEG_functions/prop_deg_function.rds')
define_behavior_degs<- readRDS('Functions/define_behavior_degs')
clown_go<- readRDS('Functions/clown_go')
define_degs<- readRDS('Functions/define_degs')

library(hdWGCNA)
library(WGCNA)

}

```

Load in the Seurat Object and Subset to Radial Glia, find subclusters
```{r}
obj <- readRDS('/Users/ggraham/Desktop/snRNA-seq R Files 122524/RNA Object.rds')
radial_glia <- subset(obj, harmony.wnn_res0.4_clusters==2)
radial_glia <- FindSubCluster(radial_glia, cluster = 2, subcluster.name = 'sub', graph.name = 'harmony.wsnn')
```

Identify the subtypes from 
https://journals.biologists.com/dev/article/147/1/dev185595/222950/Single-cell-sequencing-of-radial-glia-progeny

Find markers first
```{r}
glia_markers <- FindAllMarkers(object = radial_glia, assay = 'RNA', group.by = 'sub')
```
Make a heatmap
```{r}
top_glia_markers <- split(glia_markers,f= glia_markers$cluster)
markers <- data.frame()
for(i in unique(glia_markers$cluster)){
  data <- top_glia_markers[[i]]
  top_marker_gene <- data$gene[data$p_val_adj == min(data$p_val_adj)]
  out_data <- data.frame(top_marker_gene = top_marker_gene,
                         cluster = i)
  markers <- rbind(markers, out_data)
}

```

```{r}
print(DotPlot(radial_glia, features=markers$top_marker_gene, group.by = 'sub')+
  coord_flip())
print(markers)
```
2_0 = latent-transforming growth factor beta-binding protein 4
2_1 = HERV-H LTR-associating protein 1 
2_2= uncharacterized LOC111568382
2_3 = uncharacterized LOC118469603
2_4 = uncharacterized LOC111576017
2_5 =

Im realizing a better way to do this is to do GO of all significant find all markers gees
```{r}
degs_2_0 <- glia_markers$gene[glia_markers$p_val_adj<0.05& glia_markers$cluster=='2_0']
go_2_0 <- clown_go(degs_2_0)
dotplot(go_2_0)

for(i in 0:9){
degs_2_1 <- glia_markers$gene[glia_markers$p_val_adj<0.05& glia_markers$cluster==paste0('2_',i)]
go_2_1 <- clown_go(degs_2_1)
print(dotplot(go_2_1)+ labs(title = paste0('2_',i)))
}


```

So it seems like 
0) progenitor?
1) migrating cells?
2) migrating cells
3) migrating cells?
4) migrating cells?
5) wtf does cell differetiation mean
6) Alright these are all the same huh
7) ooh wnt interesting
8) Same old
9) ok 2_9 is interesting, it seems a lot like mature(ing) neurons

```{r}
ensembl <- useEnsembl(biomart = "genes", 
                      dataset = "aocellaris_gene_ensembl")

biomart_ocellaris <-
  getBM(
    mart = ensembl, #working mart 
    attributes = c("entrezgene_accession"))

ensembl_rerio <- useEnsembl(biomart = "genes", 
                      dataset = "drerio_gene_ensembl")
biomart_drerio <- 
  getBM(
        mart = ensembl_rerio, #working mart 
        attributes = c('external_gene_name',
                   "aocellaris_homolog_associated_gene_name"))

joined <- biomart_ocellaris%>%
  right_join(biomart_drerio, join_by('entrezgene_accession'=='aocellaris_paralog_associated_gene_name'))

drerio_to_ocellaris <- function(gene){
  output_gene <- joined$entrezgene_accession[joined$drerio_homolog_associated_gene_name==gene]
  return(output_gene)
  }


genes <- unique(c('gfap', #marks both
           'LOC111577263', #brain aromatase
           'elavl3', #neuron marker
           'map2', #neuron marker
           'sv2a', #mature neuron marker
           drerio_to_ocellaris('nrgna'), #mature neuron marker,
           'camk2a', #mature neuron marker
           'neurod1', #neuron dev
           'bhlhe22', #neuron dev
           'dlx5a',#gabergic interneuron
           's100b',
           'id1',
          drerio_to_ocellaris('glula'),
           'tubb5',
           'cd99l2',
           drerio_to_ocellaris('ppp1r14ba'),
           'cnp',
           drerio_to_ocellaris('tuba2'),
          drerio_to_ocellaris('fabp7a'),
            drerio_to_ocellaris('slc1a2b')
           ))

DotPlot(radial_glia, features=genes, group.by = 'sub')+
  coord_flip()

```
OK, from this we can determine
0) 





